File: programmingclub/moocast/moocast.py
fin = open('moocast.in', 'r')
fout = open('moocast.out', 'w')

N = int(fin.readline())
cows = [] # array of (x, y, power)  

for i in range(N):
    line = [int(x) for x in fin.readline().split()] # read entire line into array
    cows.append(line)

adjList = [[]] * N # initialize with N empty arrays
for i in range(N):
    for j in range(N):
        if i == j:
            continue

        dx = abs(cows[i][0] - cows[j][0]) # difference in x-coordinates
        dy = abs(cows[i][1] - cows[j][1]) # difference in y-coordinates

        if dx * dx + dy * dy <= cows[i][2] * cows[i][2]: # cow i can reach cow j
            adjList[i].append(j)

visited = []
def dfs(node):
    if visited[node]:
        return
    visited[node] = True

    for i in adjList[node]:
        dfs(i)

maxcows = 1
for i in range(N):
    visited = [False] * N # initialize with N false values
    dfs(i)

    total = 0 # total cows reached
    for j in range(N):
        if visited[j]:
            total += 1
    
    maxcows = max(maxcows, total)

fout.write(str(maxcows))
fout.close()

File: programmingclub/moocast/moocast.java
import java.io.*;
import java.util.*;

public class moocast {
    static ArrayList<Integer>[] adjList;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        Scanner in = new Scanner(new File("moocast.in"));
        PrintWriter out = new PrintWriter("moocast.out");

        int N = in.nextInt();

        Cow[] cows = new Cow[N];

        for (int i = 0; i < N; i++) {
            cows[i] = new Cow(in.nextInt(), in.nextInt(), in.nextInt());
        }

        adjList = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            adjList[i] = new ArrayList<>();
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (i == j) continue;

                int dx = Math.abs(cows[i].x - cows[j].x); // difference in x-coordinates
                int dy = Math.abs(cows[i].y - cows[j].y); // difference in y-coordinates

                if (dx * dx + dy * dy <= cows[i].power * cows[i].power) { // cow i can reach cow j
                    adjList[i].add(j);
                }
            }
        }

        int max = 1;
        for (int i = 0; i < N; i++) {
            visited = new boolean[N]; // all booleans initialize to false
            dfs(i);

            int total = 0; // total cows reached
            for (int j = 0; j < N; j++) {
                if (visited[j]) total++;
            }
            max = Math.max(max, total);
        }

        out.println(max);
        out.close();
    }

    static void dfs(int node) {
        if (visited[node]) return;
        visited[node] = true;

        for (int i : adjList[node]) {
            dfs(i);
        }
    }

    static class Cow {
        int x, y, power;
        
        Cow(int x, int y, int power) {
            this.x = x;
            this.y = y;
            this.power = power;
        }
    }
}
